qiita_db/sql_connection.py

```diff
@@ -630,3 +632,381 @@ def get_temp_queue(self):
+    invoked outside a context, a RuntimeError is raised.
+    """
+
+    _regex = re.compile("^{(\d+):(\d+):(\d+)}$")
+
+    def __init__(self):
+        self._queries = []
+        self._results = []
+        self.index = 0
+        self._contexts_entered = 0
+        self._connection = None
+
+    def _open_connection(self):
+        # If the connection already exists and is not closed, don't do anything
+        if self._connection is not None and self._connection.closed == 0:
+            return
+
+        try:
+            self._connection = connect(user=qiita_config.user,
+                                       password=qiita_config.password,
+                                       database=qiita_config.database,
+                                       host=qiita_config.host,
+                                       port=qiita_config.port)
+        except OperationalError as e:
+            # catch threee known common exceptions and raise runtime errors
+            try:
+                etype = e.message.split(':')[1].split()[0]
+            except IndexError:
+                # we recieved a really unanticipated error without a colon
+                etype = ''
+            if etype == 'database':
+                etext = ('This is likely because the database `%s` has not '
+                         'been created or has been dropped.' %
+                         qiita_config.database)
+            elif etype == 'role':
+                etext = ('This is likely because the user string `%s` '
+                         'supplied in your configuration file `%s` is '
+                         'incorrect or not an authorized postgres user.' %
+                         (qiita_config.user, qiita_config.conf_fp))
+            elif etype == 'Connection':
+                etext = ('This is likely because postgres isn\'t '
+                         'running. Check that postgres is correctly '
+                         'installed and is running.')
+            else:
+                # we recieved a really unanticipated error with a colon
+                etext = ''
+            ebase = ('An OperationalError with the following message occured'
+                     '\n\n\t%s\n%s For more information, review `INSTALL.md`'
+                     ' in the Qiita installation base directory.')
+            raise RuntimeError(ebase % (e.message, etext))
+
+    def close(self):
+        if self._connection is not None:
+            self._connection.close()
+
+    @contextmanager
+    def _get_cursor(self):
+        """Returns a postgres cursor
+
+        Returns
+        -------
+        psycopg2.cursor
+            The psycopg2 cursor
+
+        Raises
+        ------
+        RuntimeError
+            if the cursor cannot be created
+        """
+        self._open_connection()
+
+        try:
+            with self._connection.cursor(cursor_factory=DictCursor) as cur:
+                yield cur
+        except PostgresError as e:
```

Comment:
Although the line 899 can propagate any other type of exception, the internal function `_execute` (where the line 899 is) is only called from the `execute` function in line 960, wrapping it in a try/except with a general Exception so it will catch everything, rollback the transaction and propagate the exception.